### PR TITLE
Job Failure

### PR DESCRIPTION
**Pull request by rustworthy**
Addresses #88

- - -
This change is [Reviewable](https://reviewable.io/reviews/jonhoo/faktory-rs/89)

### CODE REVIEW

File: src/proto/single/mod.rs [hunk 1]
```diff
@@ -206,19 +208,39 @@ impl JobBuilder {
     }
 }
 
+/// Details on a job's failure.
 #[derive(Serialize, Deserialize, Debug, Clone)]
+#[non_exhaustive]
 pub struct Failure {
-    retry_count: usize,
-    failed_at: String,
+    /// [`Number`](Job::retry) of times this job can be retried.
```
Comment (jonhoo) [inline src/proto/single/mod.rs:215]:
```suggestion
    /// [Number](Job::retry) of times this job can be retried.
```

Comment (jonhoo) [inline src/proto/single/mod.rs:215]: Are we sure that this is the number of times it _can_ be retried, as opposed to the number of times it _has_ been retried (which it feels like the name implies)

File: src/proto/single/mod.rs [hunk 2]
```diff
@@ -206,19 +208,39 @@ impl JobBuilder {
...
+    pub retry_count: usize,
+
+    /// Number of remaining retry attempts.
+    #[serde(rename = "remaining")]
+    pub retry_remaining: usize,
+
+    /// Last time this job failed.
+    pub failed_at: DateTime<Utc>,
     #[serde(skip_serializing_if = "Option::is_none")]
```
Comment (jonhoo) [inline src/proto/single/mod.rs:224]: This and the other `#[serde]` attributes seem misplaced now — they should come after the doc comments for the relevant field.

File: src/proto/single/mod.rs [hunk 3]
```diff
@@ -206,19 +208,39 @@ impl JobBuilder {
...
-    message: Option<String>,
+
+    /// Error message, if any.
+    pub message: Option<String>,
     #[serde(skip_serializing_if = "Option::is_none")]
     #[serde(rename = "errtype")]
-    kind: Option<String>,
+
+    /// Error kind, if known.
+    pub kind: Option<String>,
```
Comment (jonhoo) [inline src/proto/single/mod.rs:239]: Is kind actually an arbitrary string? Things like this is why `Failure` wasn't previously `pub`.

File: src/proto/single/mod.rs [hunk 4]
```diff
@@ -266,6 +288,17 @@ impl Job {
     pub fn failure(&self) -> &Option<Failure> {
```
Comment (jonhoo) [inline src/proto/single/mod.rs:288]: nit: this should probably return `Option<&Failure>`

File: src/proto/single/mod.rs [hunk 5]
```diff
@@ -266,6 +288,17 @@ impl Job {
...
         &self.failure
     }
+
+    /// Error message for this job, if any.
+    ///
+    /// A convenience method that, internally, will check if this
+    /// job has got a [`Failure`] and - if it does - will access
+    /// the failure's message (see [`Failure::message`]).
```
Comment (jonhoo) [inline src/proto/single/mod.rs:296]: I don't know if this convenience method is worthwhile?

File: tests/real/community.rs [hunk 6]
```diff
@@ -795,46 +800,183 @@ async fn test_jobs_with_blocking_handlers() {
...
+    let local = "test_panic_and_errors_in_handler";
+    let mut c = Client::connect().await.unwrap();
+    let pattern = format!(r#"*\"args\":\[\"{}\"\]*"#, local);
+    c.requeue(
+        MutationTarget::Retries,
+        MutationFilter::builder().pattern(pattern.as_str()).build(),
+    )
+    .await
+    .unwrap();
+    c.queue_remove(&[local]).await.unwrap();
```
Comment (jonhoo) [inline tests/real/community.rs:832]: Why is this necessary?

File: tests/real/community.rs [hunk 7]
```diff
@@ -795,46 +800,183 @@ async fn test_jobs_with_blocking_handlers() {
...
+    // ... consume all the jobs from the queue and _fail_ them
+    // "in different ways" (see our worker setup above)
+    for _ in 0..njobs {
+        assert!(w.run_one(0, &[local]).await.unwrap());
+    }
 
-    // we _did_ consume and process the job, the processing result itself though
-    // was a failure; however, a panic in the handler was "intercepted" and communicated
-    // to the Faktory server via the FAIL command;
-    // note how the test run is not interrupted with a panic
```
Comment (jonhoo) [inline tests/real/community.rs:823]: Feels worthwhile to preserve this comment in the new code.

File: tests/real/community.rs [hunk 8]
```diff
@@ -795,46 +800,183 @@ async fn test_jobs_with_blocking_handlers() {
...
+    // let's now make sure all the jobs are re-enqueued
+    c.requeue(
+        MutationTarget::Retries,
+        MutationFilter::builder().pattern(pattern.as_str()).build(),
+    )
+    .await
+    .unwrap();
+
+    // now, let's create a worker who will only send jobs to the
+    // test's main thread to make some assertions
```
Comment (jonhoo) [inline tests/real/community.rs:896]: I'd like a description here about why we have to do this in two passes. I _think_ it's because the failure information in the job isn't something that's passed back to the client, it's something that's passed "forward" to the next handler the job is passed to? But that's not spelled out here.

File: tests/real/community.rs [hunk 9]
```diff
@@ -881,11 +1026,43 @@ async fn mutation_requeue_jobs() {
         .await
         .unwrap();
 
-    // the job has been re-enqueued and we consumed it again
-    let had_one = worker.run_one(0, &[local]).await.unwrap();
-    assert!(had_one);
+    // ... this time, instead of failing the job this time, let's
+    // create a new woker that will just send the job
+    // to the test thread so that we can inspect and
+    // assert on the failure
```
Comment (jonhoo) [inline tests/real/community.rs:1032]:
```suggestion
    // assert on the failure from the first run
```

File: src/proto/single/mod.rs (hunk 3 above)
Comment (rustworthy) [reply src/proto/single/mod.rs:239]: hmm, but then why would we have `failure` accessor in our public API. 

I understand that breaking changes are unpleasant, but do you think the fact that the `kind` may become "enumerated"  in Faktory (even if this enumeration were not exhaustive)  tomorrow should prevent us from exposing `Failure` in the bindings today? Or you mean more research is needed on whether this can enumerated?

Let me double-check though

File: src/proto/single/mod.rs (hunk 5 above)
Comment (rustworthy) [reply src/proto/single/mod.rs:296]: Maybe we can nuke it, I will then use a helper function in the tests binaries

File: tests/real/community.rs (hunk 6 above)
Comment (rustworthy) [reply tests/real/community.rs:832]: This clean up logic is needed because we are  (well at least I am) re-using the same container when developing locally. Without it, Faktory will re-schedule the failed jobs _after_ the test run, and so the next test run will face this remainders

File: src/proto/single/mod.rs (hunk 2 above)
Comment (rustworthy) [reply src/proto/single/mod.rs:224]: Sorry for this, must have tilted

File: src/proto/single/mod.rs (hunk 1 above)
Comment (rustworthy) [reply src/proto/single/mod.rs]: You are right and we are actually [demonstrating](https://github.com/jonhoo/faktory-rs/blob/feature/job-failure/tests/real/community.rs#L1054-L1058) this in the tests

File: tests/real/community.rs (hunk 7 above)
Comment (rustworthy) [reply tests/real/community.rs]: threw the baby out with the bathwater

File: tests/real/community.rs (hunk 8 above)
Comment (rustworthy) [reply tests/real/community.rs]: Exactly, and it is also available via Web UI. Add t a comment:
```
we _did_ consume and process the job, the processing result itself though
was a failure; however, a panic in the handler was "intercepted" and communicated
to the Faktory server via the FAIL command, the error message and the backtrace, if any,
will then be available to a Web UI user or to the worker consumes the retried job
(if `Job::retry` is Some and > 0 and there are `Failure::retry_remaining` attempts left)
in this job's `failure` field; see how we are consuming the retried jobs later
in this test and examin the failure details;
```

File: src/proto/single/mod.rs (hunk 3 above)
Comment (rustworthy) [reply src/proto/single/mod.rs:242]: In our codebase, the kind (`errtype` in Faktory) is always an "unknown" literal, we construct a `Fail` of this kind with `Fail::generic` all the time. 

Also "unknown" is what they use in [go bindings](https://github.com/contribsys/faktory/blob/10ccc2270dc2a1c95c3583f7c291a51b0292bb62/client/client.go#L363). Notice how this is actually hard-coded.

I wish we could enumerate this (with a non-exhaustive enum where `Unknown` would be the only variant for now), but what if there are also other workers (in other languages or not using the `faktory-rs` bindings), how do we unmarshal that? 

___

The more I am thinking about it, the more I am questioning if there is any usefulness in this field at all.
We can not make `kind` public for now at all, and if in the future it is no longer a hard-coded "unknown" literal, we could consider enumerating it, wdyt?

Comment (jonhoo) [reply src/proto/single/mod.rs:242]: Yeah, I like that idea — let's just make that field not `pub` for now!

File: tests/real/community.rs (hunk 6 above)
Comment (jonhoo) [reply tests/real/community.rs:832]: Would you mind leaving a slightly more elaborate comment to that effect?

Comment (rustworthy) [reply tests/real/community.rs:832]: Sure thing

**Review by jonhoo — APPROVED** [review]
Approving subject to the last two open threads which have easy fixes :+1:
